### DISCUSSION TITLE
I2S Input working example

### DISCUSSION THREAD

**Discussion by enukic** [earlephilhower/arduino-pico#2148] [General]
Hi, I have been trying to get I2S input running using following components:
Raspberry Pi Pico board and [es8388 codec](https://pcbartists.com/products/es8388-module/esp32-es8388-audio-codec-interfacing/), and I have the following pinout:
| Pin name | ES8388    | Pico Pin No. |
| :-----: | :---: | :---: |
| MCK/MCLK| MCLK | 26 |
| BCK/SCLK| SCLK| 27|
| WS/LRCLK| LRCLK| 28 |
| Data out| DIN| 19|
| Data in| DOUT| 22|
| SCL| SCL| 21|
| SDA| SDA| 20 |

_Note: I have I2C pull-up resistors connected_

I have my code in a github repo here:
https://github.com/enukic/pico_es8388_sample

When I run the code, and if I try printing received values inside `void __not_in_flash_func(test_on_rec)(void)`
I'm only receiving zeros.

es8388 driver is based on Phil Schatzman's Arduino Audio Kit driver for the same codec. I just extracted it so I don't have the use the library in my code.

My questions are:
1. Does I2S peripheral work as input in this Core? Are there any know issues?
2. Is there anything obviously wrong with the code?
3. Any other suggestions on how can I proceed to debug this issue?

Thanks in advance!
Emir

**Comment by earlephilhower** [comment]
1. Yes, works fine and had some extensive testing, fixing one bit shift issue, in [this thread](https://github.com/earlephilhower/arduino-pico/issues/2123).
2. Few things:  Right now the GIT code doesn't have the callback uncommented, so it's not going to ever get called.  Even when called, though, you get a callback per complete buffer and not per byte.  So you need to read the whole buffer out, not 1 sample, or you'll never read through even the first buffer itself before overflow.  Also, make the number of words per buffer larger.  10 is small and will give high interrupt rate.  Unless you have some crazy latency requirements, 64 is a good start there.
3. Get a logic analyzer on the I2S input pins to verify the chip you're using is really putting out data.  If the config library isn't working and it's not sending data, you're never going to read anything back.  Also, if you can use internal MCLK instead of generating, that's 1 less pin to worry about here...

**Comment by frohro** [comment]
Hi @enukic,
Did you get this going?  I'm looking for an example taking I2S ADC data and feeding it to Phil Schatzmann's Audio Tools.
Thanks,
Rob
